### DISCUSSION TITLE
How to use the proxies from local outcome dataset

### DISCUSSION THREAD

**Discussion by mocksu** [MRCIEU/TwoSampleMR#326] [Q&A] [answered]
The question was originally presented at https://github.com/MRCIEU/TwoSampleMR/issues/253, but I think it is worth mentioning alone.

Suppose you have "SNP1" in your exposure data (named "idat"), and "SNP1" does not exist in your local outcome data. Since the outcome data is local, you cannot use "extract_outcome_data" to take care of the proxies automatically, and have to use "read_outcome_data" which does not have the proxy looking option implemented. 

Let's say you find "SNP1a" in your outcome data which is in perfect LD with "SNP1", and you put "SNP1a" in your outcome dataset (named "odat"). Now you have "SNP1" in "idat" & "SNP1a" in "odat". Then you do "harmonise_data(idat, odat)" to try to associate "SNP1" with "SNP1a". 

The problem is that it does not do the expected association of SNP1 & SNP1a. Instead, it drops "SNP1" & "SNP1a" during the harmonisation. Likewise, I am not sure how "extract_outcome_data" handles the proxies, either.

Any suggestions would be appreciated.

Thanks!

**Comment by wangyiyisheng** [comment]
Have you slove this problem？

**Reply by mocksu** [reply]
yes, i wrote a script my self to do so.

**Reply by MGjorgoska** [reply]
Can you please share it? 
I am also using read_outcome_data() from TwoSampleMR for a local GWAS summary outcome, and haven't figured out how to look for proxies within or after this function.
Thanks

**Reply by mattlee821** [reply]
this [function](https://github.com/mattlee821/functions/blob/main/R/proxy-search.R) uses the same principle as [`extract_outcome_data_internal()`](https://github.com/MRCIEU/TwoSampleMR/blob/master/R/query.R)

**Reply by MGjorgoska** [reply]
Thanks @mattlee821, that's great. 
Just checking about the local reference panel, can this be a local VCF file for dbSNP153 in eg. hg19 genomic assembly version?
Thanks

**Reply by mattlee821** [reply]
i dont think so - it uses [`get_ld_proxies()`](https://github.com/mrcieu/gwasvcf/blob/HEAD/R/proxy.r) which is set up for PLINK files - i think you would need to change the proxy search part (lines 47-55), maybe this can be done with [LDlinkR](https://github.com/CBIIT/LDlinkR?tab=readme-ov-file)...

**Reply by MGjorgoska** [reply]
Thanks @mattlee821, I was able to do it with [LDlinkR](https://github.com/CBIIT/LDlinkR?tab=readme-ov-file), works like magic :)

**Comment by TCMzhoutong** [comment]
May I ask if this issue has been resolved? Do we need to use proxy SNPs for replacement in the exposed dataset?

**Reply by laleoarrow** [reply]
I've submitted my code for this issue in a recent PR. It might help you. Please note, the current version could be unstable. Feedback is welcome. #519

**Reply by 2613df** [reply]
Thanks! And I've found some bugs.
In R/find_proxy line 52:  genome_build = gd)
Actually not gd, but gb

**Comment by mocksu** [comment, marked as answer]
Yes I've implemented a solution. Although the code can't be shared, the steps to carry out the proxy finding are as follows:

1. find a LD reference panel (e.g. 1000 Genomes), tabix it. If you are using one file for one chrom, tabix each one of them.
2. select a SNP (typically an IV SNP after pruning) that you need to find proxies for ("SNP S")
3. If "SNP S" is in your outcome dataset, you don't need to look for a proxy
4. If "SNP S" is not in your outcome dataset, find SNPs in the reference panel that have high LDs (e.g. R2=1) with "SNP S", and you get a collection of SNPs ("list P" SNPs). Since your reference panel is tabixed, the look up is fast. 
5.  If any SNP ("SNP P") in "list P" is also in your outcome dataset, use it: use rsID of "SNP S" in the place of "SNP P" and remember to change the alleles of "SNP P" to those of "SNP S". Since effect allele frequency is needed for 2sMR, you need to be careful to get the alleles correctly, otherwise the direction of the beta could be wrong during the replacement.
